frontend/src/api/adk.ts

```diff
@@ -235,6 +235,42 @@ export const extractADKText = (response: ADKResponse|undefined, onlyLast: boolea
       return textParts.map(x => x.text).join("\n");
   })
 
+  // Check for structured error responses
+  const allText = textSections.join("\n");
+  if (allText.includes("MANUGEN_ERROR:")) {
+    const errorMatch = allText.match(/MANUGEN_ERROR:\s*(\{.*?\})/s);
+    if (errorMatch) {
+      try {
+        const errorData = JSON.parse(errorMatch[1]);
+        throw new ManugenError(
+          errorData.error_type || 'unknown_error',
+          errorData.message || 'An error occurred',
+          errorData.details || '',
+          errorData.suggestion || ''
+        );
+      } catch (e) {
+        if (e instanceof ManugenError) {
+          throw e;
+        }
+        // If JSON parsing fails, throw a generic error
+        throw new ManugenError(
+          'parse_error',
+          'Failed to parse error response',
+          'The server returned an error but it could not be parsed properly.',
+          'Please try again or contact support if the problem persists.'
+        );
+      }
+    } else {
+      // MANUGEN_ERROR found but no valid JSON match
+      throw new ManugenError(
+        'parse_error',
+        'Malformed error response detected',
+        'The server returned an error but it could not be parsed properly.',
+        'Please try again or contact support if the problem persists.'
+      );
+    }
+  }
+
```

Comment:
First, I dislike it looking for a sentinel string to detect an error. This code will also erroneously report an error if "MANUGEN_ERROR:" occurs anywhere in the text. IMO if you were going to use sentinel strings for errors, you should perform this check over each `textPart` and see if it begins with `MANUGEN_ERROR:`. 

Even better, in the server-sent events case, this error-detecting code should be moved to the server-side events handler, so that errors are displayed immediately rather than once the entire response has been received.

Regarding how errors are flagged, you could add the `error_code` and `error_message` fields to the event to make it a proper ADK error; this code would then look for those fields and report them.